src/checkers/flymake-collection-statix.el

```diff
@@ -0,0 +1,134 @@
+
+;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+;; IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+;; FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+;; AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+;; LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+;; OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+;; SOFTWARE.
+
+;;; Commentary:
+
+;; This file provides a flymake backend that lets the user analyze nix code
+;; using statix.
+
+;;;; Tips
+
+;; + You can customize `flymake-collection-statix-command' to your liking.
+
+;; + Some warnings have "{fix}" suffix.  They can be fixed by calling
+;;   `flymake-collection-statix-fix'.
+
+;;;; Credits
+
+;; This diagnostic would not have been possible without the statix[1] program.
+
+;;  [1] https://github.com/nerdypepper/statix
+
+;;; Code:
+
+;;;; Requirements
+
+(require 'flymake)
+(require 'flymake-collection)
+
+(eval-when-compile
+  (require 'flymake-collection-define))
+
+;;;; Customization
+
+(defcustom flymake-collection-statix-command
+  '("statix" "check" "--format" "json" "--stdin")
+  "Statix command.
+Make sure that it accepts json on stdin."
+  :type '(repeat :tag "Command" (string :tag "Part"))
+  :group 'flymake-collection)
+
+;;;; Commands
+
+(defun flymake-collection-statix-fix ()
+  "Fix region at point using statix suggestion."
+  (interactive)
+  (when-let ((data (cl-loop for diagnostic in (flymake-diagnostics (point))
+                            for data = (flymake-diagnostic-data diagnostic)
+                            when data return data)))
+    (save-restriction
+      (widen)
+      (let-alist data
+        (delete-region .beg .end)
+        (insert .fix)))))
+
+;;;; Functions
+
+;;;;; Public
+
+;;;###autoload (autoload 'flymake-collection-statix "flymake-collection-statix")
+(flymake-collection-define-enumerate flymake-collection-statix
+  "A Nix syntax and style checker using statix.
+
+See URL `https://github.com/nerdypepper/statix'."
+  :title "statix"
+  :pre-check (unless (executable-find (car flymake-collection-statix-command))
+               (error "Cannot find statix executable"))
+  :command flymake-collection-statix-command
+  :generator
+  (let ((data (list (cons 'buf flymake-collection-source))))
+    (mapcan
+     (lambda (report)
+       (let-alist report
+         (let ((data (cons (cons 'severity .severity) data)))
+           (mapcar (apply-partially #'append data) .diagnostics))))
+     (cdr (assq 'report (car (flymake-collection-parse-json
+                              (buffer-substring-no-properties
+                               (point-min) (point-max))))))))
+  :enumerate-parser
+  (let-alist it
+    (list .buf
+          (car (flymake-diag-region .buf .at.from.line .at.from.column))
+          (car (flymake-diag-region .buf .at.to.line .at.to.column))
```

Comment:
Should this be `cdr`? Also in `flymake-collection-statix--make-data`.

Reply:
Nope, `car` element is the point at line and column